### PR TITLE
Keyword argument support

### PR DESCRIPTION
**Pull request by g-arjones**
Follow-up from: https://github.com/doudou/flexmock/pull/21

This was built on top of https://github.com/doudou/flexmock/tree/keyword_argument_support to make reviewing what was added/changed to the work that was already done easier. Feel free to change the base branch if you prefer to review the whole thing together. Also, please don't forget to change the base branch before merging.

@doudou, as you suggested, support for ruby 2.5 was dropped as it would require the `ruby2_keywords` workaround and handling positional and keyword arguments together in all methods that delegate calls (there are many of them throughout the codebase), which seemed counterproductive.  Also, a new predicate (`Expectation#with_kw_args(kw)`) had to be implemented in order to support "partially" matching keyword arguments (against `HashMatcher`).

### CODE REVIEW

**Comment by g-arjones** [comment]
I would suggest releasing this as `3.0.0`. Projects on Ruby 2.5 will have to stick to flexmock<3.

**Comment by g-arjones** [comment]
One thing to note is that the design change (separating kw and positional args in the interface) makes it, as far as I can tell, impossible to write portable test code for methods that receive any number of positional arguments and no keyword arguments if the last argument is a hash (or even worse, anything that responds to `to_hash`) because on ruby2 it will be implicitly converted to keyword arguments in the mocked interfaces.

In summary, to use flexmock to effectively test something like:

```ruby
def foo(*args)
    ...
end
```

One will have to write different test cases for ruby2 and ruby3. After porting autoproj and autobuild test suites I feel like if flexmock is supposed to keep supporting ruby2, maybe the previous implementation + ruby2_keywords would be a better solution.

**Comment by g-arjones** [comment]
@doudou Ping?

**Comment by g-arjones** [comment]
> given the breakage I'd like to give it a proper look

Definitely.

Any idea when you will be able to get back to it?

**Comment by g-arjones** [comment]
@doudou Ping?

File: lib/flexmock/expectation.rb [hunk 1]
```diff
@@ -211,6 +211,13 @@ def with_any_positional_args
       self
     end
 
+    # Declare that the method should be called with the given
+    # keyword arguments
+    def with_kw_args(kw)
```
Comment (doudou) [inline lib/flexmock/expectation.rb:216]: Wouldn't it make more sense to write this `with_kw_args(**kw)` ?

Would be weird to have to write `with_kw_args({ kw: 10 })`

**Comment by doudou** [comment]
Listen. I deeply apologize for stalling this this long ... let's say that personal life got in the way big time, but that's no sucks.

I'll trust you on this. Let's just discuss the finer point of the with_kw_args API (comment above). I'll run it against existing Rock code to see how the result looks like, and then we release 3.0. I'll just have to stick to < 3.0 for 2.7 for the time being "just to be sure".

File: lib/flexmock/expectation.rb (hunk 1 above)
Comment (g-arjones) [reply lib/flexmock/expectation.rb:216]: Keywords get converted to hash implicitly on single argument methods (this behavior has not changed on ruby3), plus, receiving as a positional argument allows matching against the special matches (AnyHash or the partial matcher...)

**Comment by g-arjones** [comment]
> Listen. I deeply apologize for stalling this this long ... let's say that personal life got in the way big time, but that's no sucks.

No need to apologize, I wasn't rushing you, just wanted to make sure you hadn't forgotten.

**Comment by doudou** [comment]
OK ... stalling again.

I'm going to merge this on master, and then I need to figure out how to publish it so that it does not break rock when on 2.7.

**Comment by doudou** [comment]
Hey ... an update.

My plan right now is to release flexmock 3.0 that would require ruby >= 3.0 only, but provide flexmock 2.4 that is compatible with ruby 2.7 *and* provides the same APIs (e.g. with_kw_args)

Right now, the two versions sit on branches 2.4-pre and 3.0-pre

The only additional change is that I've added an API to test for blocks that do not depend on adding a Proc as the last argument of `with()`

I've successfully started migrating some big Ruby packages that we're running, testing them with 3.0-pre. I need to test that 2.4-pre (1) does not break the existing test suites and (2) is indeed compatible with the tests of the ruby 3.x version.

**Comment by g-arjones** [comment]
Glad to hear that. Thank you for taking the time. Right now I have to maintain internal forks while still trying to contribute to the upstream repos which really sucks.

What are your plans for autoproj and autobuild? Do you want to keep them compatible with ruby25? [I've recently had to add conditional dependencies do autobuild's gemspec](https://github.com/rock-core/autobuild/blob/master/autobuild.gemspec) because a few gems are no longer default on newer versions and can't be installed on older versions. I believe we would be able to fix that by dropping 2.5 support.

**Comment by g-arjones** [comment]
Not as you would expect:

```
root@0ec62b81cb12:/# gem install net-smtp
Fetching: timeout-0.4.1.gem (100%)
ERROR:  Error installing net-smtp:
	The last version of timeout (>= 0) to support your Ruby & RubyGems was 0.4.0. Try installing it with `gem install timeout -v 0.4.0` and then running the current command again
	timeout requires Ruby version >= 2.6.0. The current ruby version is 2.5.0.
root@0ec62b81cb12:/# ruby --version
ruby 2.5.1p57 (2018-03-29 revision 63029) [x86_64-linux-gnu]
root@0ec62b81cb12:/# 
```

This was on a Ubuntu 18.04 docker container.

**Comment by doudou** [comment]
That's `gem`. `bundler` does (which is what autoproj, well ... the whole Ruby ecosystem uses) ...

image

**Comment by doudou** [comment]
Anyways. Just release v3.0.0 and v2.4.0

**Comment by g-arjones** [comment]
Cool :+1:
